File: Sorting/Python/mergesort.py
import math

temp = [0 for i in range(10)]

def merge(arr, l, m, r):
    i, j, k = l, m, l
    while i < m and j <= r:
        if arr[i] < arr[j]:
            temp[k] = arr[i]
            i += 1
        else:
            temp[k] = arr[j]
            j += 1
        k += 1

    while i < m:
        temp[k] = arr[i]
        i += 1
        k += 1
    while j <= r:
        temp[k] = arr[j]
        j += 1
        k += 1

    for x in range(l, r+1):
        arr[x] = temp[x]

def merge_sort(arr, l, r):
    if l<r:
        mid = math.floor((r+l)/2)
        merge_sort(arr, l, mid)
        merge_sort(arr, mid+1, r)
        merge(arr, l, mid+1, r)
        

arr = [12, 17, 2, -1, 0, 1, 15, 100, 8, 19]
print(arr)
merge_sort(arr, 0, 9)
print(arr)

File: Sorting/C++/mergesort.cpp
#include<iostream>
#include<cmath>

#define ARR_SIZE 10

using namespace std;

void Merge(int a[], int l, int m, int r){
    int i = l, j = m, k = l;
    int b[r-l+1];
    while(i<m && j<=r){
        if(a[i]<a[j]){
            b[k++] = a[i++];
        } else {
            b[k++] = a[j++];
        }
    }

    for(; i<m; i++, k++){
        b[k] = a[i];
    }
    for(; j<=r; j++, k++){
        b[k] = a[j];
    }

    for(k=l; k<=r; k++){
        a[k] = b[k];
    }
}

void MergeSort(int a[], int l, int r){
    if(l<r){
        int mid = floor((float)(l+r)/2);
        MergeSort(a, l, mid);
        MergeSort(a, mid+1, r);
        Merge(a, l, mid+1, r);
    }
}


void printArray(int a[], int n){
    for(int i=0; i<n; i++){
        cout << a[i] << "\t";
    }
    cout << endl;
}

int main(){
    int a[ARR_SIZE] = {12, 17, 2, -1, 0, 1, 15, 100, 8, 19};

    printArray(a, ARR_SIZE);
    MergeSort(a, 0, ARR_SIZE-1);
    printArray(a, ARR_SIZE);
    
};